Bluetooth: Mesh: Enhance Provision speed over PB-ADV bearer

LingaoM opened the pull request:
https://github.com/zephyrproject-rtos/zephyr/pull/48903
https://github.com/zephyrproject-rtos/zephyr/pull/57883

LingaoM committed:
Bluetooth: Mesh: Remove bits for adv tag

since tag for buf single only, no need for bit, also for
save some memory for rfu.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Rename adv relay to adv simultaneous

Since notice that simultaneous advertising is not only used
by relay message, provision over pb-adv can also be used.
so it was changed to a more general name.

refs:https://github.com/zephyrproject-rtos/zephyr/pull/48903

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM committed:
Bluetooth: Mesh: Use system workqueue for dhkey gen

Since the default process dhkey gen in bt rx, will block send
Trans Ack, cause peer device send more package.

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM commented:
[Provision.zip](https://github.com/zephyrproject-rtos/zephyr/files/12547621/Provision.zip)

LingaoM committed:
Bluetooth: Mesh: Move ext adv sector to vector

Obviously, it looks obscure by putting it in a sector,
so, let's move to vector.

refs: https://github.com/zephyrproject-rtos/zephyr/pull/57883

Signed-off-by: Lingao Meng <menglingao@xiaomi.com>

LingaoM commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -401,17 +423,20 @@ void bt_mesh_adv_init(void)
+	for (int i = 0; i < ARRAY_SIZE(advs); i++) {
+		(void)memcpy(&advs[i].adv_param, &adv_param, sizeof(adv_param));
+
+		k_work_init_delayable(&advs[i].work, send_pending_adv);
```
too same init

LingaoM replied:
done

alxelax commented on subsys/bluetooth/mesh/prov_device.c:
```diff
@@ -342,6 +344,15 @@ static void prov_dh_key_gen(void)
+	prov_dh_key_gen();
+}
+
+static K_WORK_DEFINE(dh_gen_work, prov_dh_key_gen_handler);
```
What will happen if device/provisioner didn't succeed to generate dh pair but ack has been already sent?

PavelVPV replied:
Ack acknowledges reception of the message but not the confirmation of generation of ECDH secret. If the device failes to generate ECDH secret, it will send another Provisioning Failed PDU: https://github.com/zephyrproject-rtos/zephyr/blob/9cec6db2fab528189aac1bd96ae560ba4c7f3fb3/subsys/bluetooth/mesh/provisioner.c#L435-L437

alxelax commented:
Could this PR wait until Monday? I'd like to review it but have limited Internet acceess at the moment.

PavelVPV commented:
The spec requires to add 20-50ms random delay between each Generic Provisioning PDUs:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/677018/3ba58ff0-01fd-45cf-94e9-9cb3fa4e9df3)

and Transaction Acks:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/677018/d1f3057a-88da-4d57-81dc-1e4ce532876e)

With old implementation we could violate (it is a shell statement) the requirement for Acks. With this change we will violate both as relay sets doesn't consider 20ms interval.

Would you mind to fix this in your PR?

alxelax reviewed (changes requested):
In general, LGTM (especially splitting on tags and tag bits that confused me previously) but PR requires a bit more additional work. Please see my comments.

alxelax commented on subsys/bluetooth/mesh/adv.c:
```diff
@@ -202,29 +202,30 @@ struct net_buf *bt_mesh_adv_buf_get(k_timeout_t timeout)
-#endif
 
-	return bt_mesh_adv_buf_get(timeout);
+#if CONFIG_BT_MESH_RELAY_ADV_SETS
```
Hm.. How is it built if `CONFIG_BT_MESH_RELAY_ADV_SETS` is zero? Seems C compilation error.

LingaoM replied:
Will not compile.

LingaoM replied:
Got it.

alxelax commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -192,6 +192,38 @@ config BT_MESH_UNPROV_BEACON_INT
+
+config BT_MESH_PB_ADV_TRANS_ACK_RETRANSMIT_COUNT
+	int "Link Ack and Transaction Ack retransmit count"
+	default 0
```
It was `2` by default before changes.

LingaoM replied:
Done.

alxelax commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -192,6 +192,38 @@ config BT_MESH_UNPROV_BEACON_INT
+
+config BT_MESH_PB_ADV_LINK_CLOSE_RETRANSMIT_COUNT
+	int "Link Close retransmit count"
+	default 7 if BT_MESH_PB_ADV_USE_RELAY_SETS
```
Why are so many retransmission attempts required in the case of relay set usage? both link close and pdu sending? 
Especially pdu sending. Growing the number of attempts from 0 to 7 seems strange. Any reason for this?

LingaoM replied:
Will terminate if ack received.

LingaoM replied:
Good point. 0 to 7 seems strange for provision, due to Provision not define such as `xmit`, but for current implementation, just use mesh message `xmit` also for provision pdu sending, could change this strange method to more general ways, but not in this PR, thanks.

alxelax commented on subsys/bluetooth/mesh/provisioner.c:
```diff
@@ -20,6 +20,8 @@
 
 #include "common/bt_str.h"
 
+#include "host/long_wq.h"
```
This including is not required.

alxelax commented:
Seems much better. Just a couple of questions and minor changes.

LingaoM commented:
[log.txt](https://github.com/zephyrproject-rtos/zephyr/files/12836378/log.txt)

@alxelax @PavelVPV BBSim CI Failed, But i can't found problem, upload log for your, sorry.

LingaoM commented:
@PavelVPV Please take a look, thanks

PavelVPV commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -69,72 +69,71 @@ struct bt_mesh_ext_adv {
-#else /* !CONFIG_BT_MESH_RELAY_ADV_SETS */
-	return &adv_main;
-#endif /* CONFIG_BT_MESH_RELAY_ADV_SETS */
+	if (!!(CONFIG_BT_MESH_RELAY_ADV_SETS)) {
```
What if `CONFIG_BT_MESH_RELAY_ADV_SETS` is not defined? It can happen if BT_MESH_RELAY is not enabled:
https://github.com/zephyrproject-rtos/zephyr/blob/c938a53e1d0ce03df027351aa1c487c405d6b36c/subsys/bluetooth/mesh/Kconfig#L106-L110

LingaoM replied:
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/adv_ext.c#L32

LingaoM replied:
No problem for this.

PavelVPV reviewed (commented):
Only one comment.

LingaoM commented on subsys/bluetooth/mesh/adv.h:
```diff
@@ -26,10 +26,19 @@ enum bt_mesh_adv_type {
+	BT_MESH_ADV_TAG_PROV,
+};
+
+enum bt_mesh_adv_tags {
```
Done.

PavelVPV commented:
@LingaoM I need to look why the BabbleSim test fails. I added a 2 seconds delay before the config message that was failing and it help, but then it failed in another config message sending. It becomes suspicious as the bsim tests start failing here and there with the same reason. I'm busy with qualification work now. I'll come back to your PR ASAP.

PavelVPV committed:
bsim: bluetooth: mesh: Increase Net Transmit Count value on the node

This is to increase probability of reception of responses (Config Status
messages) from the node when the provisioner sends a Set message and the
node response with a Status message at the same time so that the message
collide.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV commented:
> @LingaoM I need to look why the BabbleSim test fails. I added a 2 seconds delay before the config message that was failing and it help, but then it failed in another config message sending. It becomes suspicious as the bsim tests start failing here and there with the same reason. I'm busy with qualification work now. I'll come back to your PR ASAP.

@LingaoM could you please cherry-pick this commit? https://github.com/zephyrproject-rtos/zephyr/commit/48227d1c82533c3e636b2637ad95f75df191d155

The issue is that the node responses with a Status message immediately after receiving a Set message and both devices start advertising at the same time. I increased Network Transmit Count state value on the node so that a Status message is always advertised more than a Set message. This should reduce probability of failure. We should and were planning to add delay with randomization for responses, which I hope will help to avoid such issues in future.

LingaoM commented:
> > @LingaoM I need to look why the BabbleSim test fails. I added a 2 seconds delay before the config message that was failing and it help, but then it failed in another config message sending. It becomes suspicious as the bsim tests start failing here and there with the same reason. I'm busy with qualification work now. I'll come back to your PR ASAP.
> 
> @LingaoM could you please cherry-pick this commit? [21cc9a5](https://github.com/zephyrproject-rtos/zephyr/commit/21cc9a5c1fb65687cab1b624134035b2e2669b21)
> 
> The issue is that the node responses with a Status message immediately after receiving a Set message and both devices start advertising at the same time. I increased Network Transmit Count state value on the node so that a Status message is always advertised more than a Set message. This should reduce probability of failure. We should and were planning to add delay with randomization for responses, which I hope will help to avoid such issues in future.
> 
> Edited: updated the commit hash

It's works, thanks!

alxelax reviewed (commented):
Just minor comment. Please change PR name. It doesn't change adv names.
